src/jsc/jsc_value.hpp

```diff
@@ -150,25 +150,47 @@ inline bool is_bson_type(JSContextRef ctx, const JSValueRef &value, std::string
         return false;
     }
 
-    jsc::String bsonTypeValue = JSValueToStringCopy(ctx, bsonType, &error);
+    jsc::String bson_type_value = JSValueToStringCopy(ctx, bson_type, &error);
     // Since the string's retain value is +2 here, we need to manually release it before returning.
-    JSStringRelease(bsonTypeValue);
+    JSStringRelease(bson_type_value);
     if (error) {
         throw jsc::Exception(ctx, error);
     }
 
-    std::string bsonTypeStringValue = bsonTypeValue;
-    return bsonTypeStringValue == type;
+    return (std::string)bson_type_value == type;
+}
+
+/**
+ * Checks if a `value` is an EJSON representation of a particular type (determined by the existance of a particular property).
+ */
+inline bool is_ejson_type(JSContextRef ctx, const JSValueRef &value, std::string property_name) {
+    if (JSValueIsNull(ctx, value) || JSValueIsUndefined(ctx, value) || !JSValueIsObject(ctx, value)) {
+        return false;
+    }
+
+    JSValueRef error = nullptr;
+    JSObjectRef object = JSValueToObject(ctx, value, &error);
+    if (error) {
+        throw jsc::Exception(ctx, error);
+    }
+
+    JSStringRef property_name_string = JSStringCreateWithUTF8CString(property_name.c_str());
+    auto property = JSObjectGetProperty(ctx, object, property_name_string, &error);
+    if (error) {
+        throw jsc::Exception(ctx, error);
+    }
+
+    return JSValueIsUndefined(ctx, property) == false;
 }
 
 template<>
 inline bool jsc::Value::is_decimal128(JSContextRef ctx, const JSValueRef &value) {
-    return is_bson_type(ctx, value, "Decimal128");
+    return is_bson_type(ctx, value, "Decimal128") || is_ejson_type(ctx, value, "$numberDecimal");
```

Comment:
This approach will execute the guards in ln. 167-175 / 133+142 twice on EJSON types, which seems unnecessary.  Is there a way to avoid this?
(Also, to reduce code duplications, consider if it would be practical to move this guards into a separate function.

Reply:
I agree this is not efficient.
Since I'm pretty much just building on the existing pattern, I would like to keep this as is and create an issue to track this for the future.